test/k8s-integration/main.go

```diff
@@ -157,69 +178,134 @@ func handle() error {
 		}
 	}
 
+	// Create temporary directories for kubernetes builds
+	k8sParentDir := getTempDir()
+	k8sDir := filepath.Join(k8sParentDir, "kubernetes")
+	testParentDir := getTempDir()
+	testDir := filepath.Join(testParentDir, "kubernetes")
+	defer removeTempDir(k8sParentDir)
+	defer removeTempDir(testParentDir)
+
+	numTasks := 4
+	errChan := make(chan error, numTasks)
+	k8sDependencyChan := make(chan bool, 1)
+
+	// If kube version is set, then download and build Kubernetes for cluster creation
+	// Otherwise, either GKE or a prebuild local K8s dir is being used
+	if len(*kubeVersion) != 0 {
+		go func() {
+			err := downloadKubernetesSource(pkgDir, k8sParentDir, *kubeVersion)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to download Kubernetes source: %v", err)
+				k8sDependencyChan <- false
+				return
+			}
+			err = buildKubernetes(k8sDir)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to build Kubernetes: %v", err)
+				k8sDependencyChan <- false
+				return
+			}
+			k8sDependencyChan <- true
+			errChan <- nil
+		}()
+	} else {
+		errChan <- nil
+		k8sDir = *localK8sDir
+		k8sDependencyChan <- true
+	}
+
+	// If test version is set, then download and build Kubernetes to run K8s tests
+	// Otherwise, either kube version is set (which implies GCE) or a local K8s dir is being used
+	if len(*testVersion) != 0 && *testVersion != *kubeVersion {
+		go func() {
+			// TODO: Build only the tests
+			err := downloadKubernetesSource(pkgDir, testParentDir, *testVersion)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to download Kubernetes source: %v", err)
+				return
+			}
+			err = buildKubernetes(testDir)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to build Kubernetes: %v", err)
+				return
+			}
+			errChan <- nil
+		}()
+	} else {
+		testDir = k8sDir
+		errChan <- nil
```

Comment:
kind of odd to me that we need to push nil errors onto the `errChan` like this. Feel like we should be able to have something watch the `errChan` for any actual errors and halt the entire test suite (and execute all defers for cleanup).

Maybe something like a watcher that watches the error channel that pushes a cancel signal to a context that all goroutines take in. Let me know if this would make it more complicated, but I feel like it's more extensible so we don't have to do an error prone counting of errChan signals as we add/remove things from the integration testing